<?xml version="1.0"?>
<launch>

  <arg name="device" default="0"/>
  <arg name="width" default="640"/>
  <arg name="height" default="480"/>
  <arg name="frame_rate" default="5"/>

  <!--node name="color" pkg="vimjay" type="color">
    <param name="width" value="$(arg width)"/>
    <param name="height" value="$(arg height)"/>
    <param name="red" value="244"/>
    <param name="green" value="244"/>
    <param name="blue" value="244"/>
  </node-->

  <node name="usb_cam" pkg="usb_cam" type="usb_cam_node"
      output="screen"
      if="false">
    <param name="image_width" value="$(arg width)"/>
    <param name="image_height" value="$(arg height)"/>
    <param name="framerate" value="$(arg frame_rate)"/>
    <param name="video_device" value="/dev/video$(arg device)" />
    <!--param name="pixel_format" value="mjpeg" /-->
    <param name="pixel_format" value="yuyv"/>
  </node>

  <group ns="usb_cam">
    <node name="iir_image" pkg="image_manip" type="iir_image_node"
        output="screen"
        if="false">
      <remap from="image_in" to="image_raw"/>
      <remap from="image_out" to="image_iir"/>
    </node>

    <!-- load several images to put into a combiner node,
      for now they have to be in same resolution and type but later need
      a node to convert them if not.
      Right now these are all 64x48-->

    <node pkg="image_publisher" type="image_publisher"
      name="load0"
      args="$(find vimjay)/data/clouds.png">
      <param name="publish_rate" value="0.5"/>
      <remap from="image_raw" to="image_0"/>
    </node>
    <group ns="load0/image_raw">
    <!-- http://answers.ros.org/question/255970/how-to-deactivate-image_transport-plugins/ -->
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
    </group>

    <node pkg="image_publisher" type="image_publisher"
      name="load1"
      args="$(find vimjay)/data/circle_aa.png" >
      <param name="publish_rate" value="0.5" />
      <!-- TODO(lucasw) this isn't working -->
      <remap from="image_raw" to="image_1"/>
    </node>
    <group ns="load1/image_raw">
    <!-- http://answers.ros.org/question/255970/how-to-deactivate-image_transport-plugins/ -->
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
    </group>

    <node pkg="image_publisher" type="image_publisher"
      name="load2"
      args="$(find vimjay)/data/gradient_rect.png" >
      <param name="publish_rate" value="0.5" />
      <remap from="image_raw" to="image_2"/>
    </node>
    <group ns="load2/image_raw">
    <!-- http://answers.ros.org/question/255970/how-to-deactivate-image_transport-plugins/ -->
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
    </group>

    <node name="combine_image" pkg="image_manip" type="iir_image_node"
        output="screen">
      <param name="use_time_sequence" value="false"/>
      <rosparam param="b_coeffs">[0.5, 0.5, 0.5, 0.5, 0.1]</rosparam>
      <remap from="image_0" to="load0/image_raw"/>
      <remap from="image_1" to="load1/image_raw"/>
      <remap from="image_2" to="load2/image_raw"/>
      <remap from="image_3" to="delayed"/>
      <remap from="image_out" to="combined"/>
      <param name="frame_rate" value="4"/>
    </node>
    <group ns="image_raw">
    <!-- http://answers.ros.org/question/255970/how-to-deactivate-image_transport-plugins/ -->
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressed'
      - 'image_transport/compressedDepth'
      - 'image_transport/theora'
    </rosparam>
    </group>

    <node name="delay" pkg="image_manip" type="image_delay_node"
        output="screen"
        if="false">
      <param name="delay" value="4.0"/>
      <param name="queue_size" value="40"/>
      <remap from="image_in" to="combined"/>
      <remap from="image_out" to="delayed"/>
    </node>

    <node name="roto" pkg="image_manip" type="roto_zoom_node"
        output="screen">
      <remap from="image_in" to="combined"/>
      <remap from="image_out" to="delayed"/>
    </node>
  </group>

  <!-- this is crashing currently

  File "/opt/ros/kinetic/lib/rqt_gui/rqt_gui", line 13, in <module>
    sys.exit(main.main())
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/rqt_gui/main.py", line 59, in main
    return super(Main, self).main(argv, standalone=standalone, plugin_argument_provider=plugin_argument_provider, plugin_manager_settings_prefix=str(hash(os.environ['ROS_PACKAGE_PATH'])))
  File "/home/lucasw/catkin_ws/src/qt_gui_core/qt_gui/src/qt_gui/main.py", line 527, in main
    perspective_manager.import_perspective_from_file(self._options.perspective_file, perspective_manager.HIDDEN_PREFIX + os.path.basename(self._options.perspective_file))
  File "/home/lucasw/catkin_ws/src/qt_gui_core/qt_gui/src/qt_gui/perspective_manager.py", line 330, in import_perspective_from_file
    self._remove_perspective(perspective_name)
  File "/home/lucasw/catkin_ws/src/qt_gui_core/qt_gui/src/qt_gui/perspective_manager.py", line 305, in _remove_perspective
    self._menu_manager.remove_item(name)
AttributeError: 'NoneType' object has no attribute 'remove_item'

  -->
  <node name="rqt_generic" pkg="rqt_gui" type="rqt_gui" 
      args="--lock-perspective --freeze-layout --hide-title --perspective-file $(find image_manip)/config/iir.perspective" 
      output="screen"
      if="false">
  </node>

</launch>
